### DISCUSSION TITLE
Unencoded slash in query parameter leaves the parameter undecoded

### DISCUSSION THREAD

**Discussion by dfioravanti** [fastapi/fastapi#10527] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
## main.py

from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root(string: str):
    return {"message": string}

## test_main.py
import urllib.parse

from starlette.testclient import TestClient

from main import app


client = TestClient(app)


def test_does_not_work():

    string = "@foo/@bar"

    response = client.get(f"/?string={urllib.parse.quote(string)}")
    assert response.json() == {"message": string}

def test_works():

    string = "@foo/@bar"

    response = client.get(f"/?string={urllib.parse.quote(string, safe='')}")
    assert response.json() == {"message": string}
```

### Description

Run the tests and the `test_does_not_work` fails with `{'message': '%40foo/%40bar'} != {'message': '@foo/@bar'}`. From what  I can tell what is happening is that if an un-encoded `/` is present in the string then the unquoting process is does not happen. 

According to the [standard](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) `/` is a valid unencoded character in a query parameter. While most clients out there will encode it, it is possible that someone does not. 

### Operating System

macOS

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

python 3.9 -> 3.11

**Comment by jgould22** [comment, marked as answer]
The test client is based on httpx, what happens if you [pass params like httpx does ](https://www.python-httpx.org/quickstart/)?

something like 

```
    response = client.get("/", params={"string":string})
```

**Reply by dfioravanti** [reply]
That works but I guess it is because it gets escaped. The problem here is that (at least from how I read the standard) one can have unescaped `/` in a query parameter. And that breaks the decoding. 

From the standard

```query       = *( pchar / "/" / "?" )

   The characters slash ("/") and question mark ("?") may represent data
   within the query component.  Beware that some older, erroneous
   implementations may not handle such data correctly when it is used as
   the base URI for relative references ([Section 5.1](https://datatracker.ietf.org/doc/html/rfc3986#section-5.1)), apparently
   because they fail to distinguish query data from path data when
   looking for hierarchical separators.  However, as query components
   are often used to carry identifying information in the form of
   "key=value" pairs and one frequently used value is a reference to
   another URI, it is sometimes better for usability to avoid percent-
   encoding those characters.
```
